xbmc/addons/AddonDatabase.cpp

```diff
@@ -124,6 +124,172 @@ void CAddonDatabase::UpdateTables(int version)
     m_pDS->exec("DROP TABLE blacklist");
     m_pDS->exec("ALTER TABLE tmp RENAME TO blacklist");
   }
+  if (version < 21)
+  {
+    m_pDS->exec("CREATE TABLE installed (id INTEGER PRIMARY KEY, addonID TEXT UNIQUE, "
+        "enabled BOOLEAN, installDate TEXT, lastUpdated TEXT, lastUsed TEXT) \n");
+
+    //Ugly hack incoming! As the addon manager isnt created yet, we need to start up our own copy
+    //cpluff to find the currently enabled addons.
+    auto cpluff = std::unique_ptr<DllLibCPluff>(new DllLibCPluff());
+    cpluff->Load();
+    cp_status_t status;
+    status = cpluff->init();
+    if (status != CP_OK)
+    {
+      CLog::Log(LOGERROR, "AddonDatabase: Upgrade failed. cp_init() returned status: %i", status);
+      return;
+    }
+
+    cp_context_t* cp_context = cpluff->create_context(&status);
+
+    status = cpluff->register_pcollection(cp_context, CSpecialProtocol::TranslatePath("special://home/addons").c_str());
+    if (status != CP_OK)
+    {
+      CLog::Log(LOGERROR, "AddonDatabase: Upgrade failed. cp_register_pcollection() returned status: %i", status);
+      return;
+    }
+
+    status = cpluff->register_pcollection(cp_context, CSpecialProtocol::TranslatePath("special://xbmc/addons").c_str());
+    if (status != CP_OK)
+    {
+      CLog::Log(LOGERROR, "AddonDatabase: Upgrade failed. cp_register_pcollection() returned status: %i", status);
+      return;
+    }
+
+    status = cpluff->register_pcollection(cp_context, CSpecialProtocol::TranslatePath("special://xbmcbin/addons").c_str());
+    if (status != CP_OK)
+    {
+      CLog::Log(LOGERROR, "AddonDatabase: Upgrade failed. cp_register_pcollection() returned status: %i", status);
+      return;
+    }
+
+    cpluff->scan_plugins(cp_context, CP_SP_UPGRADE);
+
+    std::string now = CDateTime::GetCurrentDateTime().GetAsDBDateTime();
+    BeginMultipleExecute();
+    int n;
+    cp_plugin_info_t** cp_addons = cpluff->get_plugins_info(cp_context, &status, &n);
+    for (int i = 0; i < n; ++i)
+    {
+      const char* id = cp_addons[i]->identifier;
+      m_pDS->exec(PrepareSQL("INSERT INTO installed(addonID, enabled, installDate) VALUES "
+          "('%s', NOT EXISTS (SELECT * FROM disabled WHERE addonID='%s'), '%s')", id, id, now.c_str()));
+    }
+    cpluff->release_info(cp_context, cp_addons);
+    CommitMultipleExecute();
+
+    m_pDS->exec("DROP TABLE disabled");
```

Comment:
Rather accidentally dropped it above, thanks! Planning to fix all the system stuff in a different pr as it requires a bit more than just dropping the table